### PR TITLE
#161291014 Implement receive notifcation feature

### PR DESCRIPTION
**Pull request by sulenchy**
#### What does this PR do?

Implement receive notification feature

#### Description of Task to be completed?

- Write tests for NotifcationEvent model
- Write tests for NotificationController
- Add Notification model
- Add NotificationEvent model
- Implement notification via email
- Add NotificationController
- Add opt-in and opt-out functionality
- Get all notification for a user functionality

#### How should this be manually tested?
- clone the repository, cd into it, checkout this branch feature/161291014/receive-notification
- set up your .env variables and local database
- run npm install to install required packages
- run npm start to start up the server
- use postman(or any api testing tool) to test the following routes route
- create a user using the route POST `http://localhost:3000/api/v1/users/signup` providing necessary request body.
- modify account status PATCH `http://localhost:3000/api/v1/users/:userId/account/:action` providing the token in the Authorization header as shown in screen shots below.
- the action param can either be activate, deactivate or delete.
- when you set action to delete, verify that you get an email in the address provided at signup
- click this link and verify that account is deleted by checking the db
- hit `http://localhost:3000/api/v1/users/follow/2`

#### Any background context you want to provide?

N/A

#### What are the relevant pivotal tracker stories?

[#161291014](https://www.pivotaltracker.com/story/show/161291014)

#### Screenshots (if appropriate)
screen shot 2018-11-19 at 4 20 22 pm

- New Comment Notification Screenshot
screen shot 2018-11-23 at 5 58 23 pm

- New Follow Notification Screenshot
screen shot 2018-11-23 at 5 58 02 pm

#### Questions:

N/A

### CODE REVIEW

File: server/controllers/CommentController.js [hunk 1]
```diff
@@ -36,6 +40,34 @@ class CommentController {
...
+                    subject: 'New Article Reaction',
+                    body: `You have new article reaction. 
+                    Check it out at 
+                    ${process.env.API_BASE_URL}/article/${articleId}`
+                  });
+                  // add a new entry to notification events
+                  NotificationController.addNewNotificationEvent(
+                    'article reaction',
+                    userObj.User.id, user.id,
+                    'You have a new reaction on an article you liked',
```
Comment (KvNGCzA) [inline server/controllers/CommentController.js]: I think this should say 'you have a new comment on an article you liked' because it is more descriptive.

Comment (sulenchy) [reply server/controllers/CommentController.js]: You are right. Thanks

Comment (sulenchy) [reply server/controllers/CommentController.js]: Feedback implemented @KvNGCzA

File: server/controllers/FollowController.js [hunk 2]
```diff
@@ -51,6 +58,18 @@ class FollowController {
             })
               .spread((followed, status) => {
                 if (status) {
+                  // TODO 2: Also add a new entry to notification events
+                  type = 'Follow';
+                  addNewNotificationEvent(
+                    'follows',
+                    authorId,
+                    id,
+                    'You have a new follower/unfollower',
```
Comment (augustineezinwa) [inline server/controllers/FollowController.js]: This message is not specific.
 I don't know if he or she is a followed me or unfollowed me.
Please make it more descriptive

File: server/controllers/FollowController.js [hunk 3]
```diff
@@ -60,6 +79,17 @@ class FollowController {
                 })
                   .then((unfollowed) => {
                     if (unfollowed) {
+                      type = 'Unfollow';
+                      addNewNotificationEvent(
+                        'unfollows',
+                        authorId,
+                        id,
+                        'Someone unfollowed you',
```
Comment (augustineezinwa) [inline server/controllers/FollowController.js]: This message: `Someone unfollowed you`
Can you make this message more specific. example: `Abiodun suleiman` unfollowed you

File: server/models/notificationevent.js [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+      allowNull: false,
+      type: DataTypes.BOOLEAN
+    }
+  }, {});
+  NotificationEvent.associate = (models) => {
+    // associations can be defined here
+    const { User, Notification } = models;
+    NotificationEvent.belongsTo(User, {
+      foreignKey: 'receiverId',
+      onDelet: 'CASCADE'
```
Comment (augustineezinwa) [inline server/models/notificationevent.js]: it seems you have a typo on line 21 and line 25.
Consider changing this `onDelet` to `onDelete`

Comment (sulenchy) [reply server/models/notificationevent.js]: Thanks for the feedback.

File: server/controllers/FollowController.js (hunk 2 above)
Comment (sulenchy) [reply server/controllers/FollowController.js]: Feedback implemented @augustineezinwa

File: server/controllers/CommentController.js (hunk 1 above)
Comment (sulenchy) [reply server/controllers/CommentController.js]: Feedback implemented @augustineezinwa

File: server/controllers/NotificationController.js [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+      .catch(err => res.status(500)
+        .json({
+          errors: {
+            message: [err.message]
+          }
+        }));
+  }
+
+  /**
+   * @description allow a user to opt in or out of notification
```
[the remark this answers is not in the document]
Comment (sulenchy) [reply server/controllers/NotificationController.js]: Thanks, @Darthrighteous. Feedback implemented

**Pull request merged**